subprojects/core/src/main/java/org/gradle/caching/internal/tasks/TaskOutputCacheCommandFactory.java

```diff
@@ -103,16 +125,45 @@ public BuildCacheKey getKey() {
             return new BuildCacheLoadCommand.Result<TaskOutputOriginMetadata>() {
                 @Override
                 public long getArtifactEntryCount() {
-                    return unpackResult.entries;
+                    return unpackResult.getEntries();
                 }
 
                 @Override
                 public TaskOutputOriginMetadata getMetadata() {
-                    return unpackResult.originMetadata;
+                    return unpackResult.getOriginMetadata();
                 }
             };
         }
 
+        private void updateSnapshots(ImmutableListMultimap<String, FileSnapshot> propertiesFileSnapshots) {
+            ImmutableMap<String, ResolvedTaskOutputFilePropertySpec> propertySpecsMap = Maps.uniqueIndex(outputProperties, new Function<TaskOutputFilePropertySpec, String>() {
+                @Override
+                public String apply(TaskOutputFilePropertySpec propertySpec) {
+                    return propertySpec.getPropertyName();
+                }
+            });
+            ImmutableSortedMap.Builder<String, FileCollectionSnapshot> propertySnapshotsBuilder = ImmutableSortedMap.naturalOrder();
+            for (String propertyName : propertiesFileSnapshots.keySet()) {
+                ResolvedTaskOutputFilePropertySpec property = propertySpecsMap.get(propertyName);
+                List<FileSnapshot> fileSnapshots = propertiesFileSnapshots.get(propertyName);
+                FileCollectionSnapshotBuilder builder = new FileCollectionSnapshotBuilder(UNORDERED, OutputPathNormalizationStrategy.getInstance(), stringInterner);
+                builder.visitFileTreeSnapshot(fileSnapshots);
+
+                propertySnapshotsBuilder.put(propertyName, builder.build());
+
+                if (fileSnapshots.size() == 1) {
```

Comment:
❌ This is not correct. Currently, if there is only one Java source file then this would be the only file in `destinationDir` for Java compilation. This would meant that we would register one class file in the destination directory a file snapshot but we wouldn't register the destination directory itself as a file tree via `putDirectory`.